Alternative analysis

johaneshp opened the pull request:
## 📝 Description

This PR merged the bug-send-to-analysis branch into alternative_analysis branch then added some changes:
1. Filter out irrelevant keywords (keywords must be at least the length of 3, filters out unregistered stopwords such as et al, e.g., i.e., ...) Some of relevant keywords such as 'AI' is also filtered as the side effect.
2. Fix the word count method so that it does not include the count of punctuation as one word
3. Ask user for input summary of their project.

**Closes:** #113 , #114 , #115 

---

## 🔧 Type of Change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📚 Documentation added/updated
- [ ] ✅ Test added/updated
- [] ♻️ Refactoring
- [ ] ⚡ Performance improvement

---

## 🧪 Testing

> Please describe how you tested this PR (both manually and with tests).  
> Provide instructions so we can reproduce.


- [x] Test Manually: copy path of zip files containing folders of projects.

---

## ✓ Checklist

- [x] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [x] 💬 I have commented my code where needed
- [] 📖 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [ ] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [ ] 🔗 Any dependent changes have been merged and published in downstream modules
- [ ] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile

---

## 📸 Screenshots

> If applicable, add screenshots to help explain your changes

<details>
<summary>Click to expand screenshots</summary>

<img width="829" height="586" alt="Screenshot 2025-10-24 at 10 55 56 PM" src="https://github.com/user-attachments/assets/23d902bd-8fb8-4bfd-9752-04e255d7f124" />
<img width="833" height="595" alt="Screenshot 2025-10-24 at 10 56 08 PM" src="https://github.com/user-attachments/assets/5d640c21-e0b6-4157-a9ef-b4c4b5f39c61" />
<img width="837" height="579" alt="Screenshot 2025-10-24 at 10 56 27 PM" src="https://github.com/user-attachments/assets/4f68f25c-a0a4-4b25-bf92-e77ecf784532" />

<img width="838" height="591" alt="Screenshot 2025-10-24 at 10 56 37 PM" src="https://github.com/user-attachments/assets/42870762-53a8-4b5f-a573-33182d3159d7" />
<img width="837" height="225" alt="Screenshot 2025-10-24 at 10 56 47 PM" src="https://github.com/user-attachments/assets/0606c1ab-e38c-4556-bf93-cebf9be09508" />



</details>

johaneshp committed:
fix word count, move function to alt_analyze.py, filter out words with length < 3

johaneshp committed:
Ask for summary for each project

johaneshp committed:
Add comprehensive tests for project summary feature

- Add tests for group_files_by_project() function
- Add tests for ask_user_for_project_summaries() function
- Test various folder structures and edge cases
- All 35 tests passing

johaneshp committed:
Merge bug-send-to-analysis into alternative_analysis

- Integrate project_analysis.send_to_analysis() flow
- Rename run_alt_analysis() to alternative_analysis() for consistency
- Keep improved project summary features with user input
- Keep group_files_by_project() with bucket folder support
- Resolve conflicts in main.py and alt_analyze.py

johaneshp committed:
edit based on newest pr

johaneshp committed:
minor update

johaneshp committed:
undo the change in helpers.py

johaneshp committed:
Merge branch 'main' into alternative_analysis

AdaraPutri reviewed (changes requested):
Thanks for working on this, Johanes! the word count now more accurately resembles the actual word count of the files, and I like the new feature you added that asks the user to type a summary. It makes the resume feel more customizable.

I just wanted to point out a few things:

- In `def topic_extraction(text: str, n_topics: int = 5, n_words: int = 10):
    lda = LatentDirichletAllocation(n_components=min(n_topics, 3), random_state=42)`, the parameter says n_topics=5, but it's capped with min(n_topics, 3) so it'll never go above 3. Maybe make this consistent by setting the default to 3 or removing min()
- I see that you're already using `nltk.corpus.stopwords` to clean up common short words, so we probably don't need the {3,} filter anymore in `token_pattern` as it ignores valid terms like AI, TV, US, ML, UI, etc
- Some non-topical words like "yes" are still being identified as topics, which isn't really what we are expecting. Maybe add a POS (part-of-speech) filter to keep only nouns or adjectives for topic modelling (it would ignore both 'yes' and 'e.g.' since those are tagged as an interjection and a foreign word, respectively)
<img width="376" height="113" alt="Screenshot 2025-10-25 at 11 34 45 AM" src="https://github.com/user-attachments/assets/16afec7f-a08d-4566-b31c-304a50ec29b3" />

But other than that, awesome work!

johaneshp committed:
remove cap in topic extraction, remove filter words more than 3, using POS filter, added more filter of latin abbreviations.

johaneshp commented:
Hey @AdaraPutri ! thanks for the feedback, Thanks for highlighting about the cap. I removed the cap and filter based on word length, I also added POS filter for noun and adjectives, and added filters of latin abbreviation manually just to be safe.

johaneshp committed:
added POS filtering in the topic_extraction

johaneshp committed:
Merge branch 'main' into alternative_analysis

taoTimTim reviewed (approved):
Great work, Johanes. 

All tests are passing and the code is performing nicely. Everything is clear and logical. The individual project summary is accurate and the metrics found *word count, readability, keywords, etc.) are all calculated correctly for the files I tested with. 

No issues found. Approved for merge.

AdaraPutri reviewed (approved):
Nice work, Johanes! the topics are now more accurately identified. Great job overall, this was quite a lot of new packages to implement and you did it successfully. This PR is ready to be merged

johaneshp committed:
Merge branch 'main' into alternative_analysis
